### PR TITLE
[SPARK-22187][SS] Update unsaferow format for saved state in flatMapGroupsWithState to allow timeouts with deleted state

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Currently, the group state of user-defined-type is encoded as top-level columns in the UnsafeRows stores in the state store. The timeout timestamp is also saved as (when needed) as the last top-level column. Since the group state is serialized to top-level columns, you cannot save "null" as a value of state (setting null in all the top-level columns is not equivalent). So we don't let the user set the timeout without initializing the state for a key. Based on user experience, this leads to confusion.

This PR is to change the row format such that the state is saved as nested columns. This would allow the state to be set to null, and avoid these confusing corner cases. However, queries recovering from existing checkpoint will use the previous format to maintain compatibility with existing production queries.

## How was this patch tested?
Refactored existing end-to-end tests and added new tests for explicitly testing obj-to-row conversion for both state formats.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/ObjectType.scala [hunk 1]
```diff
@@ -43,7 +43,7 @@ case class ObjectType(cls: Class[_]) extends DataType {
 
   def asNullable: DataType = this
 
-  override def simpleString: String = cls.getName
+  override def simpleString: String = s"Object[${cls.getName}]"
```
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/ObjectType.scala]: this is to make it differentiate between IntegerType and ObjectType with `int` in it. Both show up in the simpleStreaing as `int`.

**Comment by tdas** [comment]
jenkins retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/FlatMapGroupsWithStateExecHelper.scala [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    override protected def getStateRow(obj: Any): UnsafeRow = {
+      require(obj != null, "State object cannot be null")
+      super.getStateRow(obj)
+    }
+  }
+
+
+  private class StateManagerImplV2(
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/FlatMapGroupsWithStateExecHelper.scala]: Add docs explaining the state format

**Comment by tdas** [comment]
@zsxwing

**Comment by zsxwing** [comment]
LGTM again

**Pull request closed without merging**